README.md
# @mconnect/cache

- Cache data operations - in memory (simple and hash cache)

## Installation

```sh
$ npm install @mconnect/cache
```

## API - Simple Cache

- Simple cache is key-value caching for any types of values (number, string and object)
- It exposes the following functions: setCache(), getCache, deleteCache and clearCache
- Simple cache API signatures:

```js
// Coming soon

const abc = 'test';
console.log('abc: ', abc);

```

## Examples - Simple Cache

```js
// Coming soon
// 
const abc = 'test';
console.log('abc: ', abc);

```

## API - Hash Cache

```js
// Coming soon

const abc = 'test';
console.log('abc: ', abc);

```

## Examples - Hash Cache

```js
// Coming soon
// 
const abc = 'test';
console.log('abc: ', abc);

```


HISTORY.md
# Historical logs

## Version 0.2.0
- Released June 09, 2019
- Resolved the toString and function variables issues
- Used function scope variables and used JSON.stringify() to clone key-value pairs

## Version 0.1.2
- Released June 03, 2019
- Fixed the deleteCache missing cases for the cacheHash

## Version 0.1.0
- Released May 24, 2019

## Version 0.1.0
- Released May 24, 2019
- Includes simple and hash cache functions



test/helpers.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-05-22 | @Updated: 2019-05-22
 * @Company: mConnect.biz | @License: MIT
 * @Description: mc-test, helper functions
 */

function sleep( ms ) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// test handler function | preferred - elegant/flexible - for handling different/simple cases
function ok( expr, msg ) {
    if ( ! expr ) {
        throw new Error(msg);
    }
}

module.exports = {sleep, ok};


test/tests.CacheData.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-05-21 | @Updated: 2019-05-21
 * @Company: mConnect.biz | @License: MIT
 * @Description: @mconnect/cache testing, simple cache
 */

const {suite, test, before} = require('mocha');

const {cacheData} = require('../index');
const {sleep, ok} = require('./helpers');

let cacheKey,
    cacheValue,
    expiryTime = 5; // in seconds

before(() => {
    cacheKey   = {
        code: 'test',
        res : 'testing',
    };
    cacheValue = {
        firstName: "abc",
        lastName : 'xyz',
    };
});

suite('@mconnect/cacheData package Testing: ', () => {
    suite('Positive testing, default options:', () => {
        test('should set and return valid cacheValue', () => {
            const setCache = cacheData.setCache(cacheKey, cacheValue, expiryTime);
            let res;
            if (setCache) {
                res = cacheData.getCache(cacheKey)['value'];
            }
            ok(JSON.stringify(res) === JSON.stringify(cacheValue), `response should be: ${JSON.stringify(cacheValue)}`);
        });
        test('should clear cache and return null/empty value', () => {
            cacheData.clearCache();
            const res = cacheData.getCache(cacheKey);
            ok(res === null, `response should be: null`);
        });
    });
    suite('Positive testing, for set options:', () => {
        test('should set and return valid cacheValue -> before timeout/expiration)', () => {
            // set redis-cache expiry time to 5 seconds
            expiryTime     = 2;
            const setCache = cacheData.setCache(cacheKey, cacheValue, expiryTime);
            let res;
            if (setCache) {
                res = cacheData.getCache(cacheKey)['value'];
            }
            ok(JSON.stringify(res) === JSON.stringify(cacheValue), `response should be: ${JSON.stringify(cacheValue)}`);
        });
        test('should return null value after timeout/expiration', async () => {
            await sleep(3000);
            const res = cacheData.getCache(cacheKey);
            ok(res === null, `response should be: null`);
        });
    });
});


test/tests.CacheHash.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-05-21 | @Updated: 2019-06-09
 * @Company: mConnect.biz | @License: MIT
 * @Description: @mconnect/cache testing, hash cache
 */

const {suite, test, before} = require('mocha');

const {cacheHash} = require('../index');
const {sleep, ok} = require('./helpers');

let cacheKey,
    cacheValue,
    hashKey,
    hashInfo,
    expiryTime = 5; // in seconds

before(() => {
    cacheKey   = {
        code: 'test',
        res : 'testing',
    };
    cacheValue = {
        firstName: "abc",
        lastName : 'xyz',
    };

    hashKey = {
        code: 'hash-test',
        res : 'hash-testing',
    };

    hashInfo = {
        key  : hashKey,
        value: cacheValue,
    };
});

suite('@mconnect/cacheHash package Testing: ', () => {
    suite('Positive testing, default options:', () => {
        test('should set and return valid cacheValue', () => {
            const setCache = cacheHash.setCache(cacheKey, hashInfo, expiryTime);
            let res;
            if (setCache) {
                res = cacheHash.getCache(cacheKey, hashKey)['value'];
            }
            ok(JSON.stringify(res) === JSON.stringify(cacheValue), `response should be: ${JSON.stringify(cacheValue)}`);
        });
        test('should clear(by key) and return null/empty value', () => {
            cacheHash.clearCache();
            const res = cacheHash.getCache(cacheKey, hashKey);
            ok(res === null, `response should be: null`);
        });
    });
    suite('Positive testing, for set options:', () => {
        test('should set and return valid cacheValue -> before timeout/expiration)', () => {
            // set redis-cache expiry time to 5 seconds
            expiryTime     = 2;
            const setCache = cacheHash.setCache(cacheKey, hashInfo, expiryTime);
            let res;
            // console.log('before-timeout-setCache: ', setCache);
            if (setCache) {
                res = cacheHash.getCache(cacheKey, hashKey)['value'];
                // console.log('before-timeout-res: ', res);
            }
            ok(JSON.stringify(res) === JSON.stringify(cacheValue), `response should be: ${JSON.stringify(cacheValue)}`);
        });
        test('should return null value after timeout/expiration', async () => {
            await sleep(3000);
            const res = cacheHash.getCache(cacheKey, hashKey);
            // console.log('timeout-value: ', res);
            ok(res === null, `response should be: null`);
        });
        test('should set and return valid cacheValue, for different hashKey', () => {
            cacheValue     = {
                firstName: "Abi",
                lastName : 'Akindele',
            };
            cacheKey       = {test: 'new-hash-key'};
            hashInfo       = {
                key  : cacheKey,
                value: cacheValue,
            };
            const setCache = cacheHash.setCache(cacheKey, hashInfo, expiryTime);
            console.log('new-setCache: ', setCache);
            let res;
            if (setCache) {
                res = cacheHash.getCache(cacheKey, cacheKey)['value'];
                console.log('get-cache-res: ', res);
            }
            ok(JSON.stringify(res) === JSON.stringify(cacheValue), `response should be: ${JSON.stringify(cacheValue)}`);
        });
    });
});


src/cache.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-05-19 | @Updated: 2019-06-09
 * @Company: mConnect.biz | @License: MIT
 * @Description: mc-cache
 */

let mcCache = {};

module.exports = {
    setCache(key = '', value = '', expire = 300) {
        // key and value: key:string, value:string, expire:time(seconds)
        try {
            if (!key || !value) return false;

            // stringify key-params
            const mainKey    = JSON.stringify(key);
            mcCache[mainKey] = {
                'value' : value,
                'expire': Date.now() + expire * 1000,
            };
            return true
        } catch (e) {
            console.log('error setting cache: ', e);
            return false;
        }
    },

    getCache(key = '') {
        try {
            // validate params, both are required
            if (!key) return null;

            // stringify key-params
            const mainKey = JSON.stringify(key);

            // cases: valid/not-expired, expired, not available/unknown
            if (mcCache[mainKey] && mcCache[mainKey]['expire'] > Date.now()) {
                return mcCache[mainKey];
            }

            if (mcCache[mainKey]) {
                delete mcCache[mainKey];
            }

            return null;
        } catch (e) {
            console.log('error getting cache: ', e);
            return null;
        }
    },

    deleteCache(key = '') {
        try {
            // validate params
            if (!key) return false;

            // stringify key-params
            const mainKey = JSON.stringify(key);

            // cases: available/deleted, not-available/not-deleted
            if (mcCache[mainKey]) {
                delete mcCache[mainKey];
                return true;
            }

            return false;
        } catch (e) {
            console.log('error deleting cache: ', e);
            return false;
        }
    },

    clearCache() {
        try {
            mcCache = {};
            return true;
        } catch (e) {
            console.log('error clearing cache: ', e);
            return false;
        }
    },
};


src/cacheHash.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-05-19 | @Updated: 2019-06-09
 * @Company: mConnect.biz | @License: MIT
 * @Description: mc-cache-hash
 */

let mcCacheHash = {};

module.exports = {
    setCache(key = '', hash = {}, expire = 300) {
        try {
            // key and value
            let hashKey   = hash && hash.key ? hash.key : key;
            let hashValue = hash && hash.value ? hash.value : '';

            if (!key || !hashKey || !hash.value) return false;

            // stringify key-params
            const mainKey = JSON.stringify(key);
            hashKey       = JSON.stringify(hashKey);

            if (!mcCacheHash[mainKey]) {
                mcCacheHash[mainKey] = {};
            }

            mcCacheHash[mainKey][hashKey] = {
                'value' : hashValue,
                'expire': Date.now() + expire * 1000,
            };
            return true;
        } catch (e) {
            console.log('error setting cache: ', e);
            return false;
        }
    },

    getCache(key = '', hashKey = '') {
        // console.log('mc-cache-content: ', mcCacheHash);
        try {
            // validate params, both are required
            if (!key || !hashKey) return null;

            // stringify key-params
            const mainKey      = JSON.stringify(key);
            const cacheHashKey = JSON.stringify(hashKey);

            // cases: valid/not-expired, expired, not available/unknown
            if (mcCacheHash[mainKey] && mcCacheHash[mainKey][cacheHashKey] && mcCacheHash[mainKey][cacheHashKey]['expire'] > Date.now()) {
                return mcCacheHash[mainKey][cacheHashKey];
            }

            if (mcCacheHash[mainKey] && mcCacheHash[mainKey][cacheHashKey]) {
                delete mcCacheHash[mainKey][cacheHashKey];
            }

            return null;
        } catch (e) {
            console.log('error getting cache: ', e);
            return null;
        }
    },

    deleteCache(key = '', hashKey = '') {
        try {
            // validate params, at least one is required
            if (!key && !hashKey) return false;

            // stringify key-params
            const mainKey      = JSON.stringify(key);
            const cacheHashKey = JSON.stringify(hashKey);

            // cases: available/deleted (by key/hashKey), not-available/not-deleted
            if (mainKey && cacheHashKey) {
                // delete by hashKey
                if (mcCacheHash[mainKey] && mcCacheHash[mainKey][cacheHashKey]) {
                    delete mcCacheHash[mainKey][cacheHashKey];
                    return true;
                }
            }

            if (mainKey) {
                // delete by mainKey
                if (mcCacheHash[mainKey]) {
                    delete mcCacheHash[mainKey];
                    return true;
                }
            }

            return false;
        } catch (e) {
            console.log('error deleting cache: ', e);
            return false;
        }
    },

    clearCache() {
        try {
            mcCacheHash = {};
            return true;
        } catch (e) {
            console.log('error clearing cache: ', e);
            return false;
        }
    },
};
